### ISSUE TITLE
CPU Memory leak in Executor Api

### ISSUE THREAD

**Issue by akhoroshev** [NVIDIA/TensorRT-LLM#2753]
Version [d93a2dde84eada06ae2339b4fb4e6432167a1cfd](https://github.com/NVIDIA/TensorRT-LLM/tree/d93a2dde84eada06ae2339b4fb4e6432167a1cfd)

Generate dataset
```bash
python3 prepare_dataset.py --tokenizer /tokenizer_dir --output datasets/cnn_dailymail.json dataset --dataset-name cnn_dailymail --dataset-split validation --dataset-config-name 3.0.0 --dataset-input-key article --dataset-prompt "Summarize the following article:" --dataset-output-key "highlights" --max-input-len 10000 --output-len-dist 300,30
```

Run benchmark for 1000000 requests (repeated)

```diff
diff --git a/benchmarks/cpp/utils/utils.cpp b/benchmarks/cpp/utils/utils.cpp
index 2f732e51..704b128c 100644
--- a/benchmarks/cpp/utils/utils.cpp
+++ b/benchmarks/cpp/utils/utils.cpp
@@ -95,6 +95,11 @@ Samples parseWorkloadJson(
         samples.emplace_back(Sample{std::move(input_ids), sample["output_len"], taskId});
     }
 
+    int i = 0;
+    while (samples.size() < maxNumSamples) {
+        samples.push_back(samples[i++]);
+    }
+
     if (samples.size() < maxNumSamples)
     {
         TLLM_LOG_WARNING(
```

```bash
root@d-gccore-wmcore-sc-msk01-devel:/code/tensorrt_llm/cpp/build/benchmarks# CUDA_VISIBLE_DEVICES=0,1,2,3 mpirun -n 4 ./gptManagerBenchmark --engine_dir /data/models/trt_17_tmp/4-gpu --dataset /code/tensorrt_llm/benchmarks/cpp/datasets/cnn_dailymail.json --concurrency 32 --warm_up 0 --log_level info --max_num_samples 1000000
```

Builded model config
<details>
{
    "version": "0.17.0.dev2024121700",
    "pretrained_config": {
        "mlp_bias": false,
        "attn_bias": false,
        "rotary_base": 500000.0,
        "rotary_scaling": {
            "factor": 16.0,
            "high_freq_factor": 4.0,
            "low_freq_factor": 1.0,
            "original_max_position_embeddings": 8192,
            "rope_type": "llama3"
        },
        "residual_mlp": false,
        "disable_weight_only_quant_plugin": false,
        "moe": {
            "num_experts": 0,
            "shared_expert_intermediate_size": 0,
            "top_k": 0,
            "normalization_mode": null,
            "sparse_mixer_epsilon": 0.01,
            "tp_mode": 0,
            "device_limited_n_group": 0,
            "device_limited_topk_group": 0,
            "device_limited_routed_scaling_factor": 1.0
        },
        "remove_duplicated_kv_heads": false,
        "fc_after_embed": false,
        "use_input_layernorm_in_first_layer": true,
        "use_last_layernorm": true,
        "layer_idx_offset": 0,
        "architecture": "LlamaForCausalLM",
        "dtype": "bfloat16",
        "vocab_size": 128256,
        "hidden_size": 8192,
        "num_hidden_layers": 92,
        "num_attention_heads": 64,
        "hidden_act": "silu",
        "logits_dtype": "float32",
        "norm_epsilon": 1e-05,
        "runtime_defaults": null,
        "position_embedding_type": "rope_gpt_neox",
        "num_key_value_heads": 8,
        "intermediate_size": 32768,
        "max_position_embeddings": 131072,
        "mapping": {
            "world_size": 4,
            "gpus_per_node": 8,
            "cp_size": 1,
            "tp_size": 4,
            "pp_size": 1,
            "moe_tp_size": 4,
            "moe_ep_size": 1
        },
        "quantization": {
            "quant_algo": null,
            "kv_cache_quant_algo": null,
            "group_size": 128,
            "smoothquant_val": 0.5,
            "clamp_val": null,
            "use_meta_recipe": false,
            "has_zero_point": false,
            "pre_quant_scale": false,
            "exclude_modules": null
        },
        "use_parallel_embedding": false,
        "embedding_sharding_dim": 0,
        "head_size": 128,
        "qk_layernorm": false,
        "rotary_embedding_dim": 128,
        "tie_word_embeddings": false
    },
    "build_config": {
        "max_input_len": 130048,
        "max_seq_len": 131072,
        "opt_batch_size": 8,
        "max_batch_size": 256,
        "max_beam_width": 1,
        "max_num_tokens": 4096,
        "opt_num_tokens": 256,
        "max_prompt_embedding_table_size": 0,
        "kv_cache_type": "PAGED",
        "gather_context_logits": false,
        "gather_generation_logits": false,
        "strongly_typed": true,
        "force_num_profiles": null,
        "profiling_verbosity": "layer_names_only",
        "enable_debug_output": false,
        "max_draft_len": 0,
        "speculative_decoding_mode": 1,
        "use_refit": false,
        "input_timing_cache": null,
        "output_timing_cache": "model.cache",
        "lora_config": {
            "lora_dir": [],
            "lora_ckpt_source": "hf",
            "max_lora_rank": 64,
            "lora_target_modules": [],
            "trtllm_modules_to_hf_modules": {}
        },
        "auto_parallel_config": {
            "world_size": 1,
            "gpus_per_node": 8,
            "cluster_key": "H100-PCIe",
            "cluster_info": null,
            "sharding_cost_model": "alpha_beta",
            "comm_cost_model": "alpha_beta",
            "enable_pipeline_parallelism": false,
            "enable_shard_unbalanced_shape": false,
            "enable_shard_dynamic_shape": false,
            "enable_reduce_scatter": true,
            "builder_flags": null,
            "debug_mode": false,
            "infer_shape": true,
            "validation_mode": false,
            "same_buffer_io": {
                "past_key_value_(\\d+)": "present_key_value_\\1"
            },
            "same_spec_io": {},
            "sharded_io_allowlist": [
                "past_key_value_\\d+",
                "present_key_value_\\d*"
            ],
            "fill_weights": false,
            "parallel_config_cache": null,
            "profile_cache": null,
            "dump_path": null,
            "debug_outputs": []
        },
        "weight_sparsity": false,
        "weight_streaming": false,
        "plugin_config": {
            "dtype": "bfloat16",
            "bert_attention_plugin": "auto",
            "gpt_attention_plugin": "auto",
            "gemm_plugin": "bfloat16",
            "gemm_swiglu_plugin": null,
            "fp8_rowwise_gemm_plugin": null,
            "qserve_gemm_plugin": null,
            "identity_plugin": null,
            "nccl_plugin": "bfloat16",
            "lora_plugin": null,
            "weight_only_groupwise_quant_matmul_plugin": null,
            "weight_only_quant_matmul_plugin": null,
            "smooth_quant_plugins": true,
            "smooth_quant_gemm_plugin": null,
            "layernorm_quantization_plugin": null,
            "rmsnorm_quantization_plugin": null,
            "quantize_per_token_plugin": false,
            "quantize_tensor_plugin": false,
            "moe_plugin": "auto",
            "mamba_conv1d_plugin": "auto",
            "low_latency_gemm_plugin": null,
            "low_latency_gemm_swiglu_plugin": null,
            "context_fmha": true,
            "bert_context_fmha_fp32_acc": false,
            "paged_kv_cache": true,
            "remove_input_padding": true,
            "reduce_fusion": false,
            "user_buffer": false,
            "tokens_per_block": 64,
            "use_paged_context_fmha": true,
            "use_fp8_context_fmha": false,
            "multiple_profiles": false,
            "paged_state": false,
            "streamingllm": false,
            "manage_weights": false,
            "use_fused_mlp": true,
            "pp_reduce_scatter": false
        },
        "use_strip_plan": false,
        "max_encoder_input_len": 1024,
        "use_fused_mlp": true,
        "monitor_memory": false,
        "use_mrope": false
    }
}
</details>

<img width="661" alt="Image" src="https://github.com/user-attachments/assets/aca07683-32e1-46c7-9bd3-b4fa73fa499b" />

<img width="661" alt="Image" src="https://github.com/user-attachments/assets/7470adbc-067f-4ac0-a3d3-abed287af1df" />

**This is not a problem with the last commit, because I have seen this behavior before.**

**Comment by akhoroshev** [comment]
I did some analysis using Intel Vtune while running the test for a small number of queries (400).

Below are some conclusions:

1. The end2end program has no leaks, all memory is cleared before the program terminates:

![Image](https://github.com/user-attachments/assets/0258467b-ca28-488a-84cb-554eab25025e)

2. Let's look at the time range after the model is successfully loaded and before the benchmark is completed (150s-650s)

![Image](https://github.com/user-attachments/assets/259578a4-ecec-407b-8da5-ddb42b1f4a8b)

2.1. AllReduce plugin consumes more and more memory during runtime:

![Image](https://github.com/user-attachments/assets/3764d8fc-a2f3-45e9-8483-9319719b6266)

I can't find any place in AllReduce::enqueue where the leak occurs, maybe it's the cuda runtime?

2.2. CublasLt  consumes more and more memory during runtime:

![Image](https://github.com/user-attachments/assets/c57d67da-205b-4b67-bab8-55375e810373)

I can't analyze cublas sources because it is closed.

**Comment by akhoroshev** [comment]
NCCL_VER="2.23.4-1+cuda12.6" has memory leak https://github.com/NVIDIA/nccl/issues/1534
